dask_ml/model_selection/_hyperband.py

```diff
@@ -0,0 +1,176 @@
+        return top_results
+    return []
+
+
+def _create_model(model, params):
+    model = clone(model).set_params(**params)
+    return model, {'iterations': 0}
+
+
+def _bracket_name(s):
+    return "bracket={s}".format(s=s)
+
+
+def _model_id(s, n_i):
+    return "bracket={s}-{n_i}".format(s=s, n_i=n_i)
+
+
+def _hyperband(client, model, params, x, y, R=9, eta=3, async=None,
+               dry_run=False, fit_params={}):
+    N, R, brackets = _get_nr(R, eta=eta)
+    params = iter(ParameterSampler(params, n_iter=sum(N)))
+    model_futures = {_model_id(s, n_i): client.submit(_create_model, model,
+                                                      next(params))
+                     for s, n, r in zip(brackets, N, R) for n_i in range(n)}
+
+    info = {s: [{"partial_fit_calls": r, "bracket": _bracket_name(s),
+                 "num_models": n, 'bracket_iter': 0,
+                 "model_id": _model_id(s, n_i),
+                 "iterations": 0}
+                for n_i in range(n)]
+            for s, n, r in zip(brackets, N, R)}
+
+    kwargs = {'dry_run': dry_run, 'fit_params': fit_params}
+    model_meta_futures = {meta["model_id"]:
+                          client.submit(_partial_fit,
+                                        model_futures[meta["model_id"]], x, y,
+                                        meta=meta, **kwargs)
+                          for _, r_metas in info.items()
+                          for meta in r_metas}
+
+    score_futures = [client.submit(_score, model_meta_future, x, y,
+                                   seed=hash(_id), dry_run=dry_run)
+                     for _id, model_meta_future in model_meta_futures.items()]
+
+    completed_jobs = {}
+    seq = as_completed(score_futures)
+    for future in seq:
+        result = future.result()
+
+        # TODO if we want to map partial_fit/score to blocks
+        #  i = get_xy_block_id(result)
+        #  print(result)
+
+        completed_jobs[result['model_id']] = result
+        jobs = _to_promote(result, completed_jobs.values(), eta=eta)
+        for job in jobs:
+            # This block prevents communication of the model
+            model_future = model_futures[job["model_id"]]
+            model_future = client.submit(_partial_fit, model_future, x, y,
+                                         meta=job, dry_run=dry_run,
+                                         fit_params=fit_params)
+
+            score_future = client.submit(_score, model_future, x, y,
+                                         seed=hash(job['model_id']),
+                                         dry_run=dry_run)
+            model_futures[job["model_id"]] = model_future
+            seq.add(score_future)
+
+    return list(completed_jobs.values())
+
+
+class HyperbandCV(DaskBaseSearchCV):
+    def __init__(self, model, params, max_iter=81, client=None, async=False):
+        self.model = model
+        self.params = params
+        self.client = default_client()
```

Comment:
For better or worse, scikit-learn doesn't really like code to run in the `__init__`. It'd be best to just set `self.client = client`, and then check for `client = client or default_client()` when you actually go to use it.

Reply:
Whoops, the keyword arg `client` was leftover. Is it still kosher to have `default_client()` in `__init__`? Otherwise I'd have to set it in every other function :/